src/bp/core/services/converse.ts

```diff
@@ -0,0 +1,147 @@
+import { InvalidParameterError } from 'errors'
+import { EventEmitter2 } from 'eventemitter2'
+import { inject, injectable, postConstruct } from 'inversify'
+import _ from 'lodash'
+
+import { Event } from '../sdk/impl'
+
+import { EventEngine } from './middleware/event-engine'
+
+export const converseApiEvents = new EventEmitter2()
+
+@injectable()
+export class ConverseService {
+  private readonly timeoutInMs = 2000
+  private jsonMap: Map<string, object> = new Map()
+
+  constructor(
+    @inject(TYPES.EventEngine) private eventEngine: EventEngine,
+    @inject(TYPES.UserRepository) private userRepository: UserRepository
+  ) {}
+
+  @postConstruct()
+  init() {
+    this.eventEngine.register({
+      name: 'converse.capture.payload',
+      description: 'Captures the response payload for the Converse API',
+      order: 10000,
+      direction: 'outgoing',
+      handler: (event, next) => {
+        this._handleCapturePayload(event)
+        next()
+      }
+    })
+    this.eventEngine.register({
+      name: 'converse.capture.nlu',
+      description: 'Captures the nlu output for the Converse API',
+      order: 10000,
+      direction: 'incoming',
+      handler: (event, next) => {
+        this._handleCaptureNlu(event)
+        next()
+      }
+    })
+  }
+
+  public async sendMessage(botId: string, userId: string, payload): Promise<any> {
+    if (!payload.text || !_.isString(payload.text) || payload.text.length > 360) {
+      throw new InvalidParameterError('Text must be a valid string of less than 360 chars')
+    }
+
+    await this.userRepository.getOrCreate('api', userId)
+
+    const incomingEvent = Event({
+      type: 'text',
+      channel: 'api',
+      direction: 'incoming',
+      payload,
+      target: userId,
+      botId
+    })
+
+    const timeoutPromise = this._createTimeoutPromise(userId)
+    const donePromise = this._createDonePromise(userId)
+
+    await this.eventEngine.sendEvent(incomingEvent)
+
+    return Promise.race([timeoutPromise, donePromise])
+  }
+
+  private async _createDonePromise(userId) {
+    return new Promise((resolve, reject) => {
+      const doneEvent = `done.${userId}`
+
+      converseApiEvents.removeAllListeners(doneEvent)
+      converseApiEvents.once(doneEvent, event => {
+        if (this.jsonMap.has(event.target)) {
+          const json = this.jsonMap.get(event.target)
+          this.jsonMap.delete(event.target)
+
+          return resolve(json)
+        } else {
+          return reject(`No responses found for event target "${event.target}".`)
+        }
+      })
+    })
+  }
+
+  private async _createTimeoutPromise(userId) {
+    return new Promise((resolve, reject) => {
+      const actionEvent = `action.${userId}`
+      const wait = setTimeout(() => {
+        converseApiEvents.removeAllListeners(`action.${userId}`)
+        converseApiEvents.removeAllListeners(`done.${userId}`)
+        this.jsonMap.delete(userId)
+        reject('Request timed out.')
```

Comment:
rejecting with a string is deprecated / not good practice, better create an Error instance instead